### DISCUSSION TITLE
How to model a CVRP where vehicles have multipurpose capacities for different products that cannot be mixed?

### DISCUSSION THREAD

**Discussion by FernandoDeAlmeidaAc** [google/or-tools#3718] [Routing (and legacy CP) questions]
I'm traying to model a CVRP where each vehicle has multiple tanks (capacities). These tanks can carry different types of products, but just one at a time.
Example:
1. There are three types of products (A, B, C).
2. Each node demands a most 2 of them.
3. Each vehicle has two tanks, and they can carry any product, but they cannot be mixed.

How I model this problem using OR-Tools Routing Library?

If each vehicle has one multipurpose capacity and each node has only one demand type, I can just impose routes carrying one type of product. However, it doesn't work with multiple capacities and demands.

**Comment by Mizux** [comment]
2) Do you mean a node may need `AA` or `BB` aka two products identicals or a node can also requiere `AB` thus need to be delivered by two vehicles ?

3) You mean both tanks must carry the same product or be empty ? aka `AA`, `A_` is legal but `AB` is forbidden ?

My 2 cents, supposing you model this problem with 3 capacity dimensions `load_A`, `load_B`, `load_C`
then at any node your capacity dimensions (i.e. vehicle load) have the following constraint: `sum(load_A, load_B, load_C) == max(load_A, load_B, load_C)` ?

**Reply by FernandoDeAlmeidaAc** [reply]
Hi. Thanks for answering. 

2. Let's assume each vehicle has two tanks of capacity T.  All nodes have the demands Da (for product A), Db (for product B) and Dc (for product C) so that:
   Da, Db, Dc >= 0,
   Da + Db + Dc <= 2 x T
   Da x Db x Dc = 0          # At most two types are greater than zero

3. A_, B_, AA, BB, AB, BA are all legal.

A vehicle can do more visits if it still has capacity. If a tank is half full, the vehicle can visit another node to load more of the same product into that tank.
